### PR TITLE
sum: check length of identity hash

### PR DESCRIPTION
**Pull request by schomatis**
Towards https://github.com/ipfs/go-ipfs/issues/5019.

[Edit by kevina to avoid using words that will automatically close a bug in a different repo as this won't fix the bug until it a version with this fix is published and the gx dependency is updated in `go-ipfs`)

### CODE REVIEW

File: sum.go [hunk 1]
```diff
@@ -116,8 +119,13 @@ func sumBlake2b(size uint8, data []byte) []byte {
 	return hasher.Sum(nil)[:]
 }
 
-func sumID(data []byte) []byte {
-	return data
+func sumID(data []byte, length int) ([]byte, error) {
+	if length != len(data) {
```
Comment (kevina) [inline sum.go]: this should be `if length >= 0 && length != len(data)`.

Comment (schomatis) [reply sum.go]: You're right, thanks for catching this.

File: sum.go [hunk 2]
```diff
@@ -52,7 +52,10 @@ func Sum(data []byte, code uint64, length int) (Multihash, error) {
 	default:
 		switch code {
 		case ID:
-			d = sumID(data)
+			d, err = sumID(data, length)
+			if err != nil {
+				return nil, err
+			}
```
Comment (kevina) [inline sum.go]: The `err != nil` check is not needed here as it is checked just after the `switch`.

**Comment by schomatis** [comment]
@kevina No problem, done, thanks.

**Comment by Kubuxu** [comment]
Will this catch something that this check would not: https://github.com/multiformats/go-multihash/blob/master/multihash.go#L255 ?

**Comment by schomatis** [comment]
@Kubuxu Could you elaborate some more on your question please? I'm new to the Multihash format.

From what I'm seeing in the code you're referencing, `dm.Length` is the length of the encoded digest, not the the length of the data from which the digest is derived (which is the check we're applying here).

**Comment by Kubuxu** [comment]
@schomatis can you add some tests for it

**Comment by kevina** [comment]
@Kubuxu I am confused on your concern with bitswap.  This check is only for the id hash where the digest _is the same_ as the data.  This is catching an attempt to create an invalid id hash.  If this check was not here the id hash would be truncated to the requested length in the same `Sum` function (https://github.com/multiformats/go-multihash/pull/77/files#diff-fdefc6f9af68c3ad90fb83e7df447213L93).  Truncating the hash to force the length to be consistent with the digest length is correct for most hashes, but not the id-hash, as it is a special case.

The check you point to is simply checking for an inconsistency that `Sum` will not create for reasons mentioned above.

**Comment by Kubuxu** [comment]
@kevina I wrote it in context of https://github.com/ipfs/go-ipfs/issues/5019.
For some reason, at first, I forgot that bitswap will send identity multihash as well as bare data.

**Comment by kevina** [comment]
@Kubuxu I am still confused, or you okay with the change?

**Comment by schomatis** [comment]
> @schomatis can you add some tests for it

@Kubuxu Yes, what would be the best way to add a test like that?

[`sum_test.go`](https://github.com/multiformats/go-multihash/blob/master/sum_test.go#L19) seems to only test "positive" hashes, I mean it tests hashes that will work and it doesn't test for errors (like the check we're adding here), should I add a separate test case? or should I expand the `SumTestCase` structure?

**Comment by Kubuxu** [comment]
I would add separate `Test...` for that.

**Comment by schomatis** [comment]
Sorry for the delay, added the test.

**Pull request merged**